src/nbd_input.ml

```diff
@@ -0,0 +1,57 @@
+module F = Vhd_format.F.From_file(Vhd_format_lwt.IO)
+
+open Lwt.Infix
+
+type extent = {
+  flags : int32;
+  length : int64;
+} [@@deriving rpc]
+
+(* The flags returned for the base:allocation NBD metadata context are defined here:
+   https://github.com/NetworkBlockDevice/nbd/blob/extension-blockstatus/doc/proto.md#baseallocation-metadata-context *)
+let flag_hole = 1l
+let flag_zero = 2l
+type extent_list = extent list [@@deriving rpc]
+
+let get_extents_json ~server ~export_name =
+  let extent_reader = "/opt/xensource/libexec/get_nbd_extents.py" in
```

Comment:
Couldn't we use https://github.com/xapi-project/nbd? It supports Lwt and you wouldn't need to spawn another process. Might have to add support for block status API query though.

Reply:
Yep, ideally we'll implement all these things in that library. But right now it's missing many things, like structured replies etc.

Reply:
I've created a ticket for this work: CP-27887